# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-10-28T09:18:48.96540138+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfdl50b27/model/ with prefix b32285a8f6124338
[INFO 2023-10-28T09:18:48.976411216+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:18:48.976747169+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2023-10-28T09:18:53.74014106+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8uosh5wi/model/ with prefix 2f461835d14d4609
[INFO 2023-10-28T09:18:53.752669604+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2023-10-28T09:18:53.752750725+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmp7wwo3wrt as temporary training directory


[INFO 2023-10-28T09:21:07.863073465+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7wwo3wrt/model/ with prefix b38d85b015bb4634
[INFO 2023-10-28T09:21:07.878567191+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2023-10-28T09:21:07.878654924+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2023-10-28T09:21:07.878694649+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2023-10-28T09:21:08.977135541+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxtsw1gbf/model/ with prefix d4405ce57540462b
[INFO 2023-10-28T09:21:08.982031787+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2023-10-28T09:21:11.218291896+00:00 kernel.cc:1214] Loading model from path /tmp/tmprt_b7uts/model/ with prefix d8c946ab134a46f8
[INFO 2023-10-28T09:21:11.239964849+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2023-10-28T09:21:12.387133332+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmpk9ih7x/model/ with prefix 486e95e4538742f8
[INFO 2023-10-28T09:21:12.39187564+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2023-10-28T09:21:14.206285311+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcbm9qe5p/model/ with prefix 1972d71b175840ac
[INFO 2023-10-28T09:21:14.234244807+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2023-10-28T09:21:15.401059296+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1fhdotqg/model/ with prefix 9092510fb2034243
[INFO 2023-10-28T09:21:15.407392113+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2023-10-28T09:21:16.488282592+00:00 kernel.cc:1214] Loading model from path /tmp/tmpje95l6wx/model/ with prefix 61118509f40940ff
[INFO 2023-10-28T09:21:16.491712543+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2023-10-28T09:21:18.490579758+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo8i3ck82/model/ with prefix 277d802dbf9b4640
[INFO 2023-10-28T09:21:18.499184349+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:21:18.499229262+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2023-10-28T09:21:20.117565528+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7k715b1l/model/ with prefix dd1887d2cd0948da
[INFO 2023-10-28T09:21:20.140502643+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2023-10-28T09:21:21.413495867+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqrhfizq5/model/ with prefix 230d9dcf11fc4601
[INFO 2023-10-28T09:21:21.4242136+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2023-10-28T09:21:23.812350302+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgd75nmf8/model/ with prefix cdbf31f1df0344ca
[INFO 2023-10-28T09:21:23.83072321+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2023-10-28T09:21:24.997335681+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1oy9ng5m/model/ with prefix 47b03eaa851444c5
[INFO 2023-10-28T09:21:25.003965112+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2023-10-28T09:21:26.4040926+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6g7jp_gx/model/ with prefix 6a7ed70e77474c17
[INFO 2023-10-28T09:21:26.41976287+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2023-10-28T09:21:27.595367695+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ms0a896/model/ with prefix b8d3427184574823
[INFO 2023-10-28T09:21:27.60191087+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2023-10-28T09:21:29.522674542+00:00 kernel.cc:1214] Loading model from path /tmp/tmpx1epitty/model/ with prefix 5f2a1a784ab644e6
[INFO 2023-10-28T09:21:29.534565359+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:21:29.534609329+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2023-10-28T09:21:31.10923261+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp05i1ehr/model/ with prefix df2a437e70654ba5
[INFO 2023-10-28T09:21:31.116210754+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2023-10-28T09:21:32.288526918+00:00 kernel.cc:1214] Loading model from path /tmp/tmpafkof4gj/model/ with prefix d4fbf3a6bf684c94
[INFO 2023-10-28T09:21:32.296753199+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2023-10-28T09:21:33.659737328+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvb4jshp3/model/ with prefix c4313d2c5f6a467c
[INFO 2023-10-28T09:21:33.674589197+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2023-10-28T09:21:35.069816323+00:00 kernel.cc:1214] Loading model from path /tmp/tmpynk3zze2/model/ with prefix 63a40dbd41214ed0
[INFO 2023-10-28T09:21:35.084179346+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2023-10-28T09:21:36.795485619+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb9q8s0gq/model/ with prefix ca5d59f2261e432e
[INFO 2023-10-28T09:21:36.808247215+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2023-10-28T09:21:38.720829236+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2czkgx77/model/ with prefix a0382adfa9b14829
[INFO 2023-10-28T09:21:38.740190363+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2023-10-28T09:21:40.229022976+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuyyr8zbg/model/ with prefix a8411f5d0b794997
[INFO 2023-10-28T09:21:40.245720767+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:21:40.245787471+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2023-10-28T09:21:41.675553541+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgxtbapg2/model/ with prefix 9ce6373725a74bbc
[INFO 2023-10-28T09:21:41.681279533+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2023-10-28T09:21:42.818216044+00:00 kernel.cc:1214] Loading model from path /tmp/tmpg2gy307d/model/ with prefix 6daa11f4d7e64e5f
[INFO 2023-10-28T09:21:42.824507437+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2023-10-28T09:21:44.060402024+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjaa23z1h/model/ with prefix 87932794d31b45d7
[INFO 2023-10-28T09:21:44.070329992+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2023-10-28T09:21:45.199366491+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgm236ujj/model/ with prefix 38c6eb8f60b24780
[INFO 2023-10-28T09:21:45.205641077+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2023-10-28T09:21:46.507779624+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3csx24im/model/ with prefix f6d0049f59734605
[INFO 2023-10-28T09:21:46.52095644+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2023-10-28T09:21:48.286748557+00:00 kernel.cc:1214] Loading model from path /tmp/tmptojp3fl0/model/ with prefix f8b362cacc7e4775
[INFO 2023-10-28T09:21:48.297278575+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2023-10-28T09:21:49.47957375+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgual8wxd/model/ with prefix 9a37e6b492ca474b
[INFO 2023-10-28T09:21:49.486983067+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2023-10-28T09:21:50.61618893+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvi0lmzxb/model/ with prefix 948d6149178a4237
[INFO 2023-10-28T09:21:50.621971482+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:21:50.622030659+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2023-10-28T09:21:52.018855706+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9k_0v1f9/model/ with prefix 07e9d0d4f9674ee4
[INFO 2023-10-28T09:21:52.033492126+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2023-10-28T09:21:53.851797047+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgbve3ber/model/ with prefix 2e2e152562724252
[INFO 2023-10-28T09:21:53.881912481+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2023-10-28T09:21:55.195383752+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdz6s8kia/model/ with prefix 475185b11a2f472d
[INFO 2023-10-28T09:21:55.205794299+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2023-10-28T09:21:57.004977093+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbgm06c7e/model/ with prefix d4ab45e268534161
[INFO 2023-10-28T09:21:57.013095301+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2023-10-28T09:21:58.446290075+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqw9jwz_u/model/ with prefix b7171d0167aa4622
[INFO 2023-10-28T09:21:58.462023514+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2023-10-28T09:21:59.748700531+00:00 kernel.cc:1214] Loading model from path /tmp/tmpab2k0ml4/model/ with prefix eabda5660ffa4653
[INFO 2023-10-28T09:21:59.759118948+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2023-10-28T09:22:00.992905786+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb0ekgtu2/model/ with prefix 25fb27e154434008
[INFO 2023-10-28T09:22:01.001792119+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:22:01.001861525+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2023-10-28T09:22:03.031787649+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk1917su5/model/ with prefix cd2783a83d21444a
[INFO 2023-10-28T09:22:03.048577009+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2023-10-28T09:22:04.731001995+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwp6y1nwv/model/ with prefix f6e61284fbad4208
[INFO 2023-10-28T09:22:04.741486681+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2023-10-28T09:22:06.162645952+00:00 kernel.cc:1214] Loading model from path /tmp/tmp16lv_npk/model/ with prefix 85a7acef27054081
[INFO 2023-10-28T09:22:06.177966343+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2023-10-28T09:22:07.588900869+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqvl2_hhh/model/ with prefix f13f61c43f6947cf
[INFO 2023-10-28T09:22:07.603257415+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2023-10-28T09:22:08.743834678+00:00 kernel.cc:1214] Loading model from path /tmp/tmpykqn_19g/model/ with prefix 302e0aa5a8cd4f3d
[INFO 2023-10-28T09:22:08.749088738+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2023-10-28T09:22:10.217648014+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgg0424yz/model/ with prefix 66c9a0064fc8453e
[INFO 2023-10-28T09:22:10.234602581+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2023-10-28T09:22:11.561193968+00:00 kernel.cc:1214] Loading model from path /tmp/tmphma592r8/model/ with prefix fdb4b5bc686d41fe
[INFO 2023-10-28T09:22:11.571810058+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:22:11.571871895+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2023-10-28T09:22:13.101636397+00:00 kernel.cc:1214] Loading model from path /tmp/tmpikscrfgb/model/ with prefix 268a00b7e7de4602
[INFO 2023-10-28T09:22:13.120007584+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2023-10-28T09:22:14.398838302+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe9c8c_z6/model/ with prefix 17fca2e8be3a4f3e
[INFO 2023-10-28T09:22:14.409948783+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2023-10-28T09:22:15.491231449+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsiqyg933/model/ with prefix ddb193c7f6434e4b
[INFO 2023-10-28T09:22:15.495031223+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2023-10-28T09:22:16.923781015+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5zd3vx6c/model/ with prefix 6a4a0bedb7b7448e
[INFO 2023-10-28T09:22:16.939552162+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2023-10-28T09:22:18.307754039+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdavhv4hz/model/ with prefix 4dbd4014b5f94ede
[INFO 2023-10-28T09:22:18.321823706+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2023-10-28T09:22:19.434784174+00:00 kernel.cc:1214] Loading model from path /tmp/tmppgpapbvn/model/ with prefix 3f51b1b515f849b7
[INFO 2023-10-28T09:22:19.439720839+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2023-10-28T09:22:20.624557984+00:00 kernel.cc:1214] Loading model from path /tmp/tmpd0ycwnwt/model/ with prefix aeddc16473024102
[INFO 2023-10-28T09:22:20.631950195+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2023-10-28T09:22:22.544199328+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4qn6s2vf/model/ with prefix 40774e5564474bb9
[INFO 2023-10-28T09:22:22.557349124+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:22:22.557465174+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2023-10-28T09:22:24.422771292+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7k5ju3cc/model/ with prefix 2143398ed9e540da
[INFO 2023-10-28T09:22:24.439726696+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2023-10-28T09:22:26.104025204+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzj71st7n/model/ with prefix 9d00abb80f464082
[INFO 2023-10-28T09:22:26.112342662+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2023-10-28T09:22:27.358199396+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp3x6z2zm/model/ with prefix 476fd56bc4d84519
[INFO 2023-10-28T09:22:27.366720208+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2023-10-28T09:22:28.457086689+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1fcohoyz/model/ with prefix a5625a1768894b47
[INFO 2023-10-28T09:22:28.460881596+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2023-10-28T09:22:29.89314109+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcs2jgh0n/model/ with prefix 63bfa70b84bf4e61
[INFO 2023-10-28T09:22:29.908915716+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2023-10-28T09:22:32.274273112+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1csv01wm/model/ with prefix b4e7b70c1d5140f0
[INFO 2023-10-28T09:22:32.288725503+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2023-10-28T09:22:33.50743161+00:00 kernel.cc:1214] Loading model from path /tmp/tmpehvmagni/model/ with prefix 718f3218862c47f2
[INFO 2023-10-28T09:22:33.512393369+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:22:33.512433608+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2023-10-28T09:22:34.734325544+00:00 kernel.cc:1214] Loading model from path /tmp/tmppux8g1u6/model/ with prefix 1424fbd82e35412c
[INFO 2023-10-28T09:22:34.742106889+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2023-10-28T09:22:36.026677244+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvlbr9ljt/model/ with prefix e3dbe75e18fa4099
[INFO 2023-10-28T09:22:36.037429183+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2023-10-28T09:22:37.375572701+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb4xpm3uh/model/ with prefix 3dc219dcdab04298
[INFO 2023-10-28T09:22:37.386417478+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2023-10-28T09:22:38.977900085+00:00 kernel.cc:1214] Loading model from path /tmp/tmpo_8cej2q/model/ with prefix 9e6f82de947b49cf
[INFO 2023-10-28T09:22:38.983296374+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2023-10-28T09:22:41.400043817+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbaiu8k52/model/ with prefix c62472dfc5e640e5
[INFO 2023-10-28T09:22:41.42856935+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2023-10-28T09:22:42.763842822+00:00 kernel.cc:1214] Loading model from path /tmp/tmp901mrroq/model/ with prefix 0a092ee4c6aa4461
[INFO 2023-10-28T09:22:42.774392442+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2023-10-28T09:22:44.001600386+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk673qz8o/model/ with prefix 769dfaf6845641c0
[INFO 2023-10-28T09:22:44.010003743+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:22:44.010052251+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2023-10-28T09:22:45.126918284+00:00 kernel.cc:1214] Loading model from path /tmp/tmptva0lj0o/model/ with prefix 62f4e6bc25bf4ccc
[INFO 2023-10-28T09:22:45.1323764+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2023-10-28T09:22:46.286950579+00:00 kernel.cc:1214] Loading model from path /tmp/tmpjubrguoe/model/ with prefix d81f8b014faa4eed
[INFO 2023-10-28T09:22:46.294029239+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2023-10-28T09:22:47.844974384+00:00 kernel.cc:1214] Loading model from path /tmp/tmphtz7kif9/model/ with prefix a568fcc19f7d44fa
[INFO 2023-10-28T09:22:47.865186471+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2023-10-28T09:22:49.185430566+00:00 kernel.cc:1214] Loading model from path /tmp/tmp38cvrdw0/model/ with prefix bc7eddc62f634757
[INFO 2023-10-28T09:22:49.19886776+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2023-10-28T09:22:50.389099139+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ythmgna/model/ with prefix 80b036c5dc5540ff
[INFO 2023-10-28T09:22:50.396143358+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2023-10-28T09:22:51.631667854+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3hemvjb3/model/ with prefix 472279dd96734d56
[INFO 2023-10-28T09:22:51.640877772+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2023-10-28T09:22:52.852380015+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3bzxunge/model/ with prefix 764bfa2198804842
[INFO 2023-10-28T09:22:52.861359475+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2023-10-28T09:22:54.35153481+00:00 kernel.cc:1214] Loading model from path /tmp/tmppv_o8of9/model/ with prefix 21019dc3e2df4dc0
[INFO 2023-10-28T09:22:54.368547853+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:22:54.368596116+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2023-10-28T09:22:55.607461037+00:00 kernel.cc:1214] Loading model from path /tmp/tmp87et5tu2/model/ with prefix 154ac1a78c2d48e6
[INFO 2023-10-28T09:22:55.615160912+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2023-10-28T09:22:57.564017805+00:00 kernel.cc:1214] Loading model from path /tmp/tmppeq061ug/model/ with prefix e9ebcd450cb24c73
[INFO 2023-10-28T09:22:57.57712248+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2023-10-28T09:22:59.23133809+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfb9fqvd6/model/ with prefix 0742115c27174c8b
[INFO 2023-10-28T09:22:59.240399982+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2023-10-28T09:23:00.376706393+00:00 kernel.cc:1214] Loading model from path /tmp/tmp478lxkgl/model/ with prefix 3388603310ad4996
[INFO 2023-10-28T09:23:00.381211811+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2023-10-28T09:23:01.534093424+00:00 kernel.cc:1214] Loading model from path /tmp/tmpj5wwieuu/model/ with prefix 85028833870d479a
[INFO 2023-10-28T09:23:01.539317525+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2023-10-28T09:23:02.720312584+00:00 kernel.cc:1214] Loading model from path /tmp/tmprtpvoa71/model/ with prefix f0e0b605cec14358
[INFO 2023-10-28T09:23:02.727975831+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2023-10-28T09:23:04.550706483+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfs6hklo4/model/ with prefix ef9d573c30494dfd
[INFO 2023-10-28T09:23:04.558207946+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:23:04.558251053+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2023-10-28T09:23:06.427190023+00:00 kernel.cc:1214] Loading model from path /tmp/tmp66uorh9f/model/ with prefix 39f26d26163a4cc1
[INFO 2023-10-28T09:23:06.438319723+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2023-10-28T09:23:08.352605187+00:00 kernel.cc:1214] Loading model from path /tmp/tmp941z9ja4/model/ with prefix 722b40ebff124822
[INFO 2023-10-28T09:23:08.364495176+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2023-10-28T09:23:09.666630058+00:00 kernel.cc:1214] Loading model from path /tmp/tmptmxs_dr2/model/ with prefix 2c27cb653d4244a2
[INFO 2023-10-28T09:23:09.677233923+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2023-10-28T09:23:10.967475979+00:00 kernel.cc:1214] Loading model from path /tmp/tmpz9dg4x_3/model/ with prefix da94f6fceeae4339
[INFO 2023-10-28T09:23:10.977922155+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2023-10-28T09:23:14.465790894+00:00 kernel.cc:1214] Loading model from path /tmp/tmphv2keyr9/model/ with prefix 92cf4e28c3874222
[INFO 2023-10-28T09:23:14.485947889+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2023-10-28T09:23:15.76268924+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2nf2s1x1/model/ with prefix 7aafefe9de654166
[INFO 2023-10-28T09:23:15.770118229+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:23:15.770164309+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2023-10-28T09:23:17.907505334+00:00 kernel.cc:1214] Loading model from path /tmp/tmphwu4qtx0/model/ with prefix 8e5c5547972b4f07
[INFO 2023-10-28T09:23:17.925517952+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2023-10-28T09:23:19.777207596+00:00 kernel.cc:1214] Loading model from path /tmp/tmpq32f_9h1/model/ with prefix 10736bb831564429
[INFO 2023-10-28T09:23:19.797154763+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2023-10-28T09:23:21.194910398+00:00 kernel.cc:1214] Loading model from path /tmp/tmps41k5afz/model/ with prefix 1a66a816ec3a4f42
[INFO 2023-10-28T09:23:21.20788385+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2023-10-28T09:23:22.768771783+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwi4ypkj9/model/ with prefix d7458152a71e4dbc
[INFO 2023-10-28T09:23:22.774485143+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2023-10-28T09:23:24.106575969+00:00 kernel.cc:1214] Loading model from path /tmp/tmpaup6mm_e/model/ with prefix 25c274f300444544
[INFO 2023-10-28T09:23:24.117846363+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2023-10-28T09:23:25.793366286+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxwzher46/model/ with prefix 8972982cec0b4cc0
[INFO 2023-10-28T09:23:25.800841105+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:23:25.800903187+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2023-10-28T09:23:27.36643382+00:00 kernel.cc:1214] Loading model from path /tmp/tmp_yrte0lg/model/ with prefix b9bd91cba8a6422b
[INFO 2023-10-28T09:23:27.385807859+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2023-10-28T09:23:28.683794077+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfenhl7ts/model/ with prefix ace69842704e4889
[INFO 2023-10-28T09:23:28.694787101+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2023-10-28T09:23:30.371444185+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3wp38yqt/model/ with prefix 2dc70cd2c751477e
[INFO 2023-10-28T09:23:30.378608221+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2023-10-28T09:23:31.653935408+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsmdg6ddt/model/ with prefix bb59f4c0705e4079
[INFO 2023-10-28T09:23:31.663219245+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2023-10-28T09:23:32.968743106+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3ugwiecp/model/ with prefix 831b0ceebdc84e1f
[INFO 2023-10-28T09:23:32.979912315+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2023-10-28T09:23:34.14735735+00:00 kernel.cc:1214] Loading model from path /tmp/tmpfm6do3tn/model/ with prefix caa93a67400a4887
[INFO 2023-10-28T09:23:34.153131662+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2023-10-28T09:23:35.404791895+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoxii76no/model/ with prefix c9bf0e169c0d4714
[INFO 2023-10-28T09:23:35.413836318+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2023-10-28T09:23:36.901169774+00:00 kernel.cc:1214] Loading model from path /tmp/tmpn_hxm6km/model/ with prefix 4f3e39233d2d4c4d
[INFO 2023-10-28T09:23:36.918123247+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2023-10-28T09:23:36.918171017+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 